src/Servers/Kestrel/Core/src/Internal/Infrastructure/PipeWriterHelpers/ConcurrentPipeWriter.cs

```diff
@@ -116,25 +118,38 @@ public override ValueTask<FlushResult> FlushAsync(CancellationToken cancellation
 
             if (flushTask.IsCompletedSuccessfully)
             {
-                if (_currentFlushTcs != null)
+                if (_currentFlush.IsCompletedSuccessfully)
                 {
-                    CompleteFlushUnsynchronized(flushTask.GetAwaiter().GetResult(), null);
+                    // If the new Flush has completed, the previous flush is likely to have completed,
+                    // if so we can complete the current Flush and return its result.
+                    CompleteFlushUnsynchronized();
                 }
 
                 return flushTask;
             }
 
-            // Use a TCS instead of something custom so it can be awaited by multiple awaiters.
-            _currentFlushTcs = new TaskCompletionSource<FlushResult>(TaskCreationOptions.RunContinuationsAsynchronously);
-            var result = new ValueTask<FlushResult>(_currentFlushTcs.Task);
+            var lastFlush = _currentFlush;
```

Comment:
No, its just used to comfirm if its the same as prior to `FlushAsyncAwaited` being called.

However... the assignment in `FlushAsyncAwaited` should be in lock else it will conflict with the assignments in the lock below

Reply:
Oh, its already in lock

Reply:
I'll put it in a VolitileRead just so its clear what's happening

Reply:
I'll put this on in a VolitileRead just so the intent is clear

Reply:
Actually it doesn't as the methods aren't called concurrently; there's just a race of `_currentFlush` with `FlushAsyncAwaited` when its in an incomplete state (as that method may also reassign it).

Will update to be more clear.